### ISSUE TITLE
1. синхронизация времени

### ISSUE THREAD

**Issue by Motavin** [inspark/vmon#1]
2. Контроллер
3. Запуск КОС
4. запускакется vmon, время синхронихируется с сервером. 
5. запускакется vmon, но время не синхронихирует, и для продолжения работы необходимо изменить файл /opt/sem-next/example/persistent.bin исправить первую строку на 20.  После чего всетаки переходит в режим инициализации.

**Comment by grekhss** [comment]
Необходимо уточнить несколько важных моментов:
1. В самом vMon для платформы KasperskyOS добавлена синхронизация времени по протоколу NTP. Это регулируется через расширенные настройки монитора: можно указать временную точку вручную или обозначить необходимость синхронизации времени через ntpd-демона. Если имеется в виду неработающая синхронизация через ntpd, то нужно больше данных: адрес сервера времени и наличие соединения между ним и контроллером, удаётся ли отправить NTP-пакет (снять трафик при помощи wireshark).
2. Есть синхронизация времени, которая осуществляется в рамках работы планировщика vPlan. Это сообщение по протоколу MQTT, которое должно приходить от центрального сервера. Если имеется в виду эта синхронизация, то опять-таки нужно больше данных: создан ли bridge между локальным брокером и брокером на сервере, какие топики синхронизируются между брокерами, соединён ли планировщик с локальным брокером, удаётся ли ему отправить запрос на синхронизацию (снять трафик при помощи wireshark), получает ли запрос от планировщика брокер на сервере, отправляется ли ответ от сервера, доходит ли ответ от сервера на локальный брокер, получает ли ответный пакет планировщик.

**Comment by SanSanichSan** [comment]
Где указать сервер NTP?

**Comment by grekhss** [comment]
Насколько я помню, он указывается в файле `/etc/ntp.conf`

**Comment by SanSanichSan** [comment]
Ну я ж не знаю откуда КОС читает)
Там по умолчанию:
server 0.debian.pool.ntp.org iburst
server 1.debian.pool.ntp.org iburst
server 2.debian.pool.ntp.org iburst
server 3.debian.pool.ntp.org iburst


[kos.boot.log](https://github.com/inspark/vmon/files/4063499/kos.boot.log)

**Comment by grekhss** [comment]
> Ну я ж не знаю откуда КОС читает)

Для того и создали обсуждение - чтобы помочь выяснить! =)
Мы пробовали ставить туда как IP-адрес, так и доменный адрес (в последнем случае, правда пришлось указать в /etc/hosts какому IP он соответствует). Кроме того, для целей ускорения старта рекомендую указать дополнительные параметры после iburst. Вот такие: "prefer minpoll 3 maxpoll 4". Это заставит демона чуть чаще спрашивать время и, соответственно, быстрее произойдёт синхронизация часов.

**Comment by SanSanichSan** [comment]
сделал
[kos.boot.log](https://github.com/inspark/vmon/files/4063585/kos.boot.log)

**Comment by grekhss** [comment]
В логах есть строка
`[   4.5600090] en0: DAD duplicate address 10.4.2.241 from 00:86:40:51:d4:11`
Видимо, где-то недалеко имеется ещё одно устройство с таким же IPv4 адресом. Нужно или на Вашем контроллере изменить сетевой адрес или на другом устройстве.

Второй момент - это падение сущности vfs_entity. Не могу понять почему это произошло, особенно, учитывая, что в первом логе падения не было. Кроме изменений в /etc/ntp.conf были ещё какие-то манипуляции? Без vfs_entity не будет работающей сети.

Третий момент - параметр <PlcSystemTime></PlcSystemTime> в файле vmon.conf. Если в нём указана конкретная временная точка (судя по логам это 1579004425), то изменения в /etc/ntp.conf будут конфликтовать с действиями демона ntpd. Предлагаю указать в этом параметре значение NTP. Вот так: <PlcSystemTime>NTP</PlcSystemTime>
В этом случае монитор будет ждать действий демона и конфликта не будет.

Четвертый момент (из первого лога) - это попытки соединиться с сервером:
```
mosquitto[12]: Connecting bridge ISS.SN (gw.inspark.ru:8883)
mosquitto[12]: Error creating bridge: System error returned in errno.
```
Судя по всему, имеет место одна из проблем: или нет физического доступа к gw.inspark.ru или KOS не понимает какому IP-адресу соответствует это доменное имя. В любом случае, предлагаю (временно!) выставить в файле /etc/hosts строку, которая явно задаст соответствие между gw.inspark.ru и конкретным IP-адресом.

**Comment by SanSanichSan** [comment]
менял  только /etc/ntp.conf и  /etc/hosts
NTP выставил
под Linux все видит, в хостах все так и было выставлено
[kos.boot.log](https://github.com/inspark/vmon/files/4063701/kos.boot.log)

Для ускорения отладки, вы можете со своего контроллера все проверить, адрес сервера и ключи у вас есть.

**Comment by grekhss** [comment]
С самого начала загрузки можете выложить лог?

**Comment by SanSanichSan** [comment]
[Архив.zip](https://github.com/inspark/vmon/files/4063759/default.zip)

**Comment by grekhss** [comment]
Можете попробовать вот такой вариант /etc/ntp.conf?

```
# You do need to talk to an NTP server or two (or three).
#server ntp.your-provider.example
#server ntp.sem-next.ru iburst prefer
server 0.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 1.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 2.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 3.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4

restrict -4 default kod limited nomodify nopeer noquery notrap
restrict -6 default kod limited nomodify nopeer noquery notrap
restrict 127.0.0.1
restrict ::1

# /etc/ntp.conf, configuration for ntpd; see ntp.conf(5) for help
driftfile /var/lib/ntp/ntp.drift

```

**Comment by SanSanichSan** [comment]
у вас отработало?

[kos.boot.log](https://github.com/inspark/vmon/files/4063985/kos.boot.log)

**Comment by grekhss** [comment]
У меня пока что нет стенда, к сожалению. Пытаюсь помочь исходя из логов, опыта и общих соображений.

**Comment by grekhss** [comment]
Давайте попробуем вот такой вариант /etc/ntp.conf

```
# You do need to talk to an NTP server or two (or three).
#server ntp.your-provider.example
#server ntp.sem-next.ru iburst prefer
#server 0.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 85.21.78.8 iburst prefer minpoll 3 maxpoll 4
server 1.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 2.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4
server 3.debian.pool.ntp.org iburst prefer minpoll 3 maxpoll 4

restrict -4 default kod limited nomodify nopeer noquery notrap
restrict -6 default kod limited nomodify nopeer noquery notrap
restrict 127.0.0.1
restrict ::1

# /etc/ntp.conf, configuration for ntpd; see ntp.conf(5) for help
driftfile /var/lib/ntp/ntp.drift
```
Если с этим не сработает синхронизация по NTP - значит нет доступа по сети. На всякий случай: KOS, как правило, ставит иной MAC-адрес сетевой карты, нежели Debian. Это может быть источником проблемы, поскольку нередко настройка сети устроена таким образом, что выход в Интернет имеют лишь устройства из чётко определённого списка.

**Comment by SanSanichSan** [comment]
у нас в сети нет списков по фильтрации МАК адресов.
адрес КОС пингуется с другой сети
(10.4.2.242) at 0:4:9f:a3:7e:80 on en6 ifscope [ethernet]

[kos.boot.log](https://github.com/inspark/vmon/files/4096237/kos.boot.log)

**Comment by grekhss** [comment]
> адрес КОС пингуется с другой сети

А с этой машины пингуется 0.debian.pool.ntp.org (он же по идее 85.21.78.8) ?

**Comment by SanSanichSan** [comment]
да, все работает. Только адрес уже 91.209.94.10

**Comment by grekhss** [comment]
Давайте попробуем исключить обращения по доменному имени, чтобы выяснить есть ли с этой частью проблема или нет. Для этого:

1. параметр PlcSystemTime в файле vmon.conf нужно вернуть к значению 1579004425 вместо NTP
2. в файле /etc/mosquitto/mosquitto.conf.bridge нужно заменить доменное имя gw.inspark.ru на 185.98.85.97. (Если с контроллера есть связь с облаком, то попытка создать bridge остановится на этапе проверки сертификата.)

**Comment by SanSanichSan** [comment]
Вернул+ выставил DHCP
[kos.boot.log](https://github.com/inspark/vmon/files/4101783/kos.boot.log)

**Comment by grekhss** [comment]
Похоже, что в адресе допущена ошибка: **s**185.98.85.97. Проверьте, п

**Comment by SanSanichSan** [comment]
я тоже это увидел, даже удалил пост...
полез проверять, а там все верно.

**Comment by SanSanichSan** [comment]
вот новый,  с IP вместо имени шлюза
[kos.boot.log](https://github.com/inspark/vmon/files/4101813/kos.boot.log)

**Comment by grekhss** [comment]
Соединение есть. Значит, проблема в доменном имени.
Давайте теперь попробуем пошагово менять настройки. Для начала вернём доменное имя в /etc/mosquitto/mosquitto.conf.bridge
Получение IP-адреса пускай остается через DHCP.

**Comment by SanSanichSan** [comment]
[kos.boot.log](https://github.com/inspark/vmon/files/4101839/kos.boot.log)

**Comment by grekhss** [comment]
В этом логе есть только одно подозрительное место - это строка 
Unable to copy file. Errorcode = 2 (No such file or directory)
Но на ход дальнейшего выполнения она не влияет. Видно, что мост с облаком построен, потому что брокер пишет в логи о сообщениях.
Можно посмотреть более полный лог? Хотелось бы увидеть что пишет vmon.

**Comment by SanSanichSan** [comment]
это полный лог, записей больше нет

**Comment by grekhss** [comment]
Из под какой операционки и каким приложением соединяетесь с контроллером? (Linux/Win10, Minicom/Kermit/Teraterm)

**Comment by SanSanichSan** [comment]
Darwin MacBook Pro 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64

screen /dev/tty.SLAB_USBtoUART 115200

**Comment by grekhss** [comment]
Я, к сожалению, не знаю как заставить screen показывать больше логов. Надо посмотреть в документации (что-нибудь типа `screen --help`). Нагуглить не получается потому что слово screen слишком часто используется в совершенно другом контексте.

**Comment by SanSanichSan** [comment]
screen показывает все что сыпется в порт, у меня к нему придирок не было) Он также есть и под linux
Похоже, что просто ни чего больше не пишет в консоль. 

Повтрюсь с вопросом, у вас выходит на коннект с облаком? Хоть у кого нибудь из 6?

**Comment by grekhss** [comment]
> screen показывает все что сыпется в порт

При определённом стечении обстоятельств порт, предназначенный для логирования, может захватить serial-приложение. Можете выложить содержимое конфига для него? (/etc/wb-mqtt-serial.copy) Для проверки предположения можно сразу отключить все устройства в этом конфиге (выставить для каждого элемента массива "ports" поле "enabled" в значение "false").

**Comment by SanSanichSan** [comment]
исправил
[kos.boot.log](https://github.com/inspark/vmon/files/4102735/kos.boot.log)

**Comment by grekhss** [comment]
Так. Теперь давайте попробуем вернуть обратно изменения в /etc/wb-mqtt-serial.copy.
Можете выложить содержимое конфига для GPIO-приложения? (файл /etc/wb-homa-gpio.copy)

**Comment by SanSanichSan** [comment]
Может я положу те, с которыми у вас все работает?

**Comment by grekhss** [comment]
Можно попробовать. Логи выложите?

**Comment by SanSanichSan** [comment]
[kos.boot.log](https://github.com/inspark/vmon/files/4116346/kos.boot.log)

[wb-homa-gpio.log](https://github.com/inspark/vmon/files/4116347/wb-homa-gpio.log)

**Comment by grekhss** [comment]
Давайте попробуем такой вариант конфига для GPIO:
```
{
  "device_name": "Discrete I/O",
  "channels": [
  ]
}
```

**Comment by SanSanichSan** [comment]
[kos.boot.log](https://github.com/inspark/vmon/files/4116519/kos.boot.log)

**Comment by grekhss** [comment]
Так. Пока оставляем конфиг для GPIO пустым.
И к этому надо добавить следующее изменение в конфиге SERIAL: поле "debug" в значение "true". (файл /etc/wb-homa-gpio.copy)

**Comment by SanSanichSan** [comment]
>  в конфиге SERIAL: поле "debug" в значение "true". (файл /etc/wb-homa-gpio.copy)
не понял в каком именно и где именно.
Давайте файлом?

**Comment by grekhss** [comment]
Выложите сюда содержимое файла /etc/wb-homa-gpio.copy. Я внесу изменения.

**Comment by SanSanichSan** [comment]
{
  "device_name": "Discrete I/O",
  "channels": [
  ]
}

**Comment by grekhss** [comment]
Так. Ошибся с именем файла. Прошу прощения.
Нужен /etc/wb-mqtt-serial.copy

**Comment by SanSanichSan** [comment]
[kos.boot.log](https://github.com/inspark/vmon/files/4117414/kos.boot.log)

**Comment by grekhss** [comment]
Так. С приложением screen точно всё в порядке. В конфиге GPIO не хватает полей "bank" для каждого GPIO, но сейчас это не имеет значения - он пока пустой.
Видимо, всё-таки проблема в настройках, касающихся vMon и/или vPlan.
Нужно проверить, что существуют все пути и файлы, необходимые для их работы. А именно:
**для монитора**
- /opt/sem-next/example/persistent.bin
- /opt/sem-next/example/etc/vmon.conf
- /opt/sem-next/example/var/inspark/vmon.conf.bak
- /opt/sem-next/example/bridge.ini
- /opt/sem-next/example/bridge.tmpl
- /opt/sem-next/example/etc/mosquitto/conf.d/bridge.conf
- /opt/sem-next/example/var/inspark/vsn
- /opt/sem-next/example/var/log/vmon

**для планировщика**
- 	opt/sem-next/example/var/inspark/log/vplan
- 	/opt/sem-next/example/etc/vplan.json
- 	/opt/sem-next/example/var/inspark/vplan.json.bak
- 	/opt/sem-next/example/var/inspark/vplan.time

**Comment by SanSanichSan** [comment]
root@inspark-ALQ4ICI:/# find /opt/sem-next/
/opt/sem-next/
/opt/sem-next/example
/opt/sem-next/example/etc
/opt/sem-next/example/etc/vmon.conf~
/opt/sem-next/example/etc/vmon.conf
/opt/sem-next/example/etc/.vmon.conf.swp
/opt/sem-next/example/etc/mosquitto
/opt/sem-next/example/etc/mosquitto/conf.d
/opt/sem-next/example/etc/mosquitto/conf.d/bridge.conf
/opt/sem-next/example/bridge.ini
/opt/sem-next/example/bridge.tmpl
/opt/sem-next/example/persistent.bin

**Comment by grekhss** [comment]
Так. Без пересборки образа не обойтись. Предлагается сделать работу монитора и планировщика более информативной: в файле vcom/vLog.cpp в методе vLog::Log( LMT mt, const string& Message ) нужно исправить строку
`#if 0
`
на
`#if 1
`
После этого собрать проект и положить полученный kos.image.stripped.bin на раздел внутренней памяти контроллера.

**Comment by grekhss** [comment]
Внёс поправку в предыдущий комментарий.

**Comment by SanSanichSan** [comment]
там нет строки 
`if DEBUG_VMON`
она есть в новых версиях, и там уже продукты вынесены из одного общего репозитория + библиотеки стали динамическими.
Для сборки наверняка, что то надо где то поменять...
С чего собирать и как?

**Comment by grekhss** [comment]
Процедура сборки достаточно подробно описана в инструкции, которая шла вместе с поставкой SDK и кода.

**Comment by SanSanichSan** [comment]
он не соберется с новой архитектурой как минимум, потому как структура каталогов изменилась.
Документация не описывает как положить новый код.

**Comment by SanSanichSan** [comment]
надо как то указать что б сначала библиотека собралась, потом все остальное...

**Comment by Motavin** [comment]
я правильно понял, что некая новая архитектура не совместима с нашей ОС?

**Comment by SanSanichSan** [comment]
совместима, там все тоже самое только библиотеку сделали не статической, а динамической

**Comment by Motavin** [comment]
Динамические библиотеки не совметимы с КОСью пока что.
Я не до конца понимаю какие баги и улучшения продукта здесь обсуждаются, если актуальная кодовая база в принципе не собирается под нашу ОС?

**Comment by SanSanichSan** [comment]
мы не обсуждаем улучшения)
Я как раз и просил собрать, показать нам, что все работает и потом уже попробовать собрать нам самим.

**Comment by Motavin** [comment]
В описании бага указан Actual Result: запускается vmon, но время не синхронизирует, и для продолжения работы необходимо изменить файл /opt/sem-next/example/persistent.bin исправить первую строку на 20. После чего все-таки переходит в режим инициализации.

На какой версии исходников из этого репозитория мы можем это воспроизвести?

**Comment by SanSanichSan** [comment]
ни на какой, это и есть новый/последний репозиторий с динамической библиотекой.

**Comment by SanSanichSan** [comment]
>  В описании бага указан Actual Result: запускается vmon, но время не синхронизирует, и для продолжения работы необходимо изменить файл /opt/sem-next/example/persistent.bin исправить первую строку на 20. После чего все-таки переходит в режим инициализации.

Это я запускал на том, что пришло в поставке от вас.

**Comment by Motavin** [comment]
Не понятно как пофиксить баги в версии исходников, котрая даже не собирается.
Я правильно помню, что в декабре был момент когда наши примеры были слиты с вашей кодовой базой и могли быть собраны под КОСь?

**Comment by SanSanichSan** [comment]
В части инициализации код не изменился. Сетевые настройки также к коду не относятся, по коайней мере не должны относиться.
